components/entities/api/src/main/java/org/phenotips/entities/internal/AbstractPrimaryEntityGroupManager.java

```diff
@@ -17,77 +17,225 @@
- * <li>all documents {@link #create() created} by this manager will have the name in the format
- * {@code <PREFIX><7 digit sequential number>}, unless {@link #getNextDocument()} is overridden, where:
- * <ul>
- * <li>the prefix is computed from the uppercase letters of the XClass name, excluding {@code Class}, e.g. for
- * {@code PhenoTips.FamilyGroupClass} the prefix will be {@code FG}; override {@link #getIdPrefix()} to change this
- * behavior</li>
- * <li>the number is a 0-padded 7 digit number, starting at {@code 0000001} and automatically incremented for each new
- * entity created</li>
- * </ul>
- * </li>
- * </ul>
+ * Base class for implementing specific entity groups, where members declare the groups that they belong to. The class
+ * deals with logical group operations and not with implementation.
+ * <p>
  *
- * @param <G> the type of groups handled by this manager
- * @param <E> the type of entities belonging to the groups handled by this manager; if more than one type of entities
- *            can be part of the groups, then a generic {@code PrimaryEntity} should be used instead
+ * @param <G> the type of the group containing the entities
+ * @param <E> the type of entities belonging to this group; if more than one type of entities can be part of the group,
+ *            then a generic {@code PrimaryEntity} should be used instead
  * @version $Id$
  * @since 1.3M2
  */
-@Unstable("New class and interface added in 1.3")
-public abstract class AbstractPrimaryEntityGroupManager<G extends PrimaryEntityGroup<E>, E extends PrimaryEntity>
-    extends AbstractPrimaryEntityManager<G> implements PrimaryEntityGroupManager<G, E>
+public abstract class AbstractPrimaryEntityGroupManager<G extends PrimaryEntity, E extends PrimaryEntity>
+    implements PrimaryEntityGroupManager<G, E>
 {
+    /** Entity type of group object. */
+    protected EntityReference groupEntityReference;
+
+    /** Manager of group primary entities. */
+    protected final PrimaryEntityManager<G> groupManager;
+
+    /** Entity type of member object. */
+    protected EntityReference memberEntityReference;
+
+    /** Manager of member primary entities. */
+    protected final PrimaryEntityManager<E> membersManager;
+
+    /** Logging helper object. */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @SuppressWarnings("unchecked")
+    protected AbstractPrimaryEntityGroupManager(
+        EntityReference groupEntityReference, EntityReference memberEntityReference)
+    {
+        this.groupEntityReference = groupEntityReference;
+        this.memberEntityReference = memberEntityReference;
+
+        this.membersManager = (PrimaryEntityManager<E>) getManager(memberEntityReference);
+        this.groupManager = (PrimaryEntityManager<G>) getManager(groupEntityReference);
+    }
+
+    private PrimaryEntityManager<PrimaryEntity> getManager(EntityReference entityReference)
+    {
```

Comment:
This implementation fails to find `PrimaryEntityManager`s with a type specified in angle brackets, e.g. `PrimaryEntityManager<Family>`. [Here](https://github.com/Gene42/phenotips-niaid/commit/bf58b9a65e7d3443d66951a349f003d99e0f722b#commitcomment-19613681), I needed to specify the role explicitly as the class without the type.

Can this be improved to find `PrimaryEntityManager`s with a type specified in angle brackets?